### PR TITLE
DR-879 Separate build workflows

### PR DESCRIPTION
**Pull request by Pira-nhs**
## Description

PR to separate the existing build and deploy workflow and Create New Application Build Workflow.

## Context

Required in order to be able to store artefacts onto a s3 bucket as part of the build and be able to retrieve during deploy.

## Type of changes

- [ ] Refactoring (non-breaking change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [x] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [x] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [x] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: docs/adr/assets/ADR-003/examples/golang/go.sum [hunk 1]
```diff
@@ -3,8 +3,8 @@ github.com/go-resty/resty/v2 v2.7.0/go.mod h1:9PWDzw47qPphMRFfhsyk0NnSgvluHcljSM
 github.com/golang-jwt/jwt v3.2.2+incompatible h1:IfV12K8xAKAnZqdXVzCZ+TOjboZ2keLg81eXfW3O+oY=
 github.com/golang-jwt/jwt v3.2.2+incompatible/go.mod h1:8pz2t5EyA70fFQQSrl6XZXzqecmYZeUEB8OUGHkxJ+I=
 golang.org/x/net v0.0.0-20211029224645-99673261e6eb/go.mod h1:9nx3DQGgdP8bBQD5qxJ1jj9UTztislL4KSBs9R2vV5Y=
-golang.org/x/net v0.18.0 h1:pVaXccu2ozPjCXewfr1S7xza/zcXTity9cCdXQYSjIM=
-golang.org/x/net v0.18.0/go.mod h1:NxSsAGuq816PNPmqtQdLE42eU2Fs7NoRIZrHJAlaCOE=
+golang.org/x/net v0.19.0 h1:pVaXccu2ozPjCXewfr1S7xza/zcXTity9cCdXQYSjIM=
```
[the remark this answers is not in the document]
Comment (Pira-nhs) [reply docs/adr/assets/ADR-003/examples/golang/go.sum:6]: yes it was a search all on v0.18 and bumped upto v0.19.

File: docs/adr/assets/ADR-003/examples/golang/go.sum [hunk 2]
```diff
@@ -3,8 +3,8 @@ github.com/go-resty/resty/v2 v2.7.0/go.mod h1:9PWDzw47qPphMRFfhsyk0NnSgvluHcljSM
...
+golang.org/x/net v0.19.0/go.mod h1:NxSsAGuq816PNPmqtQdLE42eU2Fs7NoRIZrHJAlaCOE=
```
[the remark this answers is not in the document]
Comment (Pira-nhs) [reply docs/adr/assets/ADR-003/examples/golang/go.sum:7]: reverted to v0.18

File: .github/actions/derive-artefact-bucket/action.yaml [hunk 3]
```diff
@@ -0,0 +1,28 @@
+name: "Derive bucket name for Push artefact action"
```
[the remark this answers is not in the document]
Comment (Pira-nhs) [reply .github/actions/derive-artefact-bucket/action.yaml:1]: amended

File: scripts/workflow/push-artefact.sh [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+# fail on first error
+set -e
+
+export ARTEFACT_BUCKET_NAME="${ARTEFACT_BUCKET_NAME:-""}"
+export WORKSPACE="${WORKSPACE:-""}"
+export COMMIT_HASH="${COMMIT_HASH:-""}"
+export DIRECTORY="${DIRECTORY:-""}"
+export SERVICE="${SERVICE:-""}"
+
```
[the remark this answers is not in the document]
Comment (Pira-nhs) [reply scripts/workflow/push-artefact.sh:11]: Agreed checks set for DIRECTORY and SERVICE

File: scripts/workflow/push-artefact.sh [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+if [ -z "$ARTEFACT_BUCKET_NAME" ] ; then
+  echo Set ARTEFACT BUCKET NAME to the artefact bucket name to action the terraform in
+  EXPORTS_SET=1
+fi
+
+echo "Deploying to S3 bucket: $ARTEFACT_BUCKET_NAME"
+echo "Workspace: $WORKSPACE"
+echo "Commit Hash: $COMMIT_HASH"
+echo "Current working directory: $(pwd)"
```
[the remark this answers is not in the document]
Comment (Pira-nhs) [reply scripts/workflow/push-artefact.sh:34]: Good catch removed

File: scripts/workflow/push-artefact.sh [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+
+# Deploy artefacts to s3 bucket
```
[the remark this answers is not in the document]
Comment (Pira-nhs) [reply scripts/workflow/push-artefact.sh:37]: amended

File: scripts/workflow/push-artefact.sh [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+  cd ./"${DIRECTORY}"/"${SERVICE}"
+  aws s3 cp ${SERVICE}.zip s3://$ARTEFACT_BUCKET_NAME/$WORKSPACE/$COMMIT_HASH/
+
+  echo "Deployment completed"
```
[the remark this answers is not in the document]
Comment (Pira-nhs) [reply scripts/workflow/push-artefact.sh:42]: amended to push

File: scripts/workflow/push-artefact.sh [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  EXPORTS_SET=1
+fi
+
+if [ -z "$ARTEFACT_BUCKET_NAME" ] ; then
+  echo Set ARTEFACT BUCKET NAME to the artefact bucket name to action the terraform in
+  EXPORTS_SET=1
+fi
+
+if [ -z "$DIRECTORY" ] ; then
+  echo Set COMMIT_HASH to identify which artefacts to be deployed
```
Comment (timrickwood) [inline scripts/workflow/push-artefact.sh:32]: cut and paste error - should be DIRECTORY

File: scripts/workflow/push-artefact.sh [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  EXPORTS_SET=1
+fi
+
+if [ -z "$SERVICE" ] ; then
+  echo Set COMMIT_HASH to identify which artefacts to be deployed
```
Comment (timrickwood) [inline scripts/workflow/push-artefact.sh:37]: should be SERVICE

File: scripts/workflow/push-artefact.sh (hunk 8 above)
Comment (Pira-nhs) [reply scripts/workflow/push-artefact.sh]: amended

File: scripts/workflow/push-artefact.sh (hunk 9 above)
Comment (Pira-nhs) [reply scripts/workflow/push-artefact.sh]: amended

**Pull request merged**